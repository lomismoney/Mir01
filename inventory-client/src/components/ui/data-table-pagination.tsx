import {
  ChevronLeftIcon,
  ChevronRightIcon,
  ChevronsLeftIcon,
  ChevronsRightIcon,
} from "lucide-react";
import { Table } from "@tanstack/react-table";

import { Button } from "@/components/ui/button";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";

/**
 * DataTable 分頁組件 - 分段進軍終章版本
 *
 * 功能特性：
 * 1. 與 TanStack Table 完全集成
 * 2. 支援手動分頁模式（後端分頁）
 * 3. 頁面大小調整功能
 * 4. 完整的分頁導航控制
 * 5. 響應式設計，適配各種螢幕尺寸
 * 6. 無障礙性支援（ARIA 標籤）
 *
 * @param table - TanStack Table 實例
 */
interface DataTablePaginationProps<TData> {
  table: Table<TData>;
  /** 可選的總數據量顯示 */
  totalCount?: number;
  /** 可選的頁面大小選項 */
  pageSizeOptions?: number[];
}

export function DataTablePagination<TData>({
  table,
  totalCount,
  pageSizeOptions = [10, 15, 20, 30, 50, 100],
}: DataTablePaginationProps<TData>) {
  const currentPage = table.getState().pagination.pageIndex + 1;
  const pageSize = table.getState().pagination.pageSize;
  const totalPages = table.getPageCount();

  // 計算顯示的數據範圍
  const startRow = table.getState().pagination.pageIndex * pageSize + 1;
  const endRow = Math.min(
    startRow + pageSize - 1,
    totalCount || table.getFilteredRowModel().rows.length,
  );
  const totalRows = totalCount || table.getFilteredRowModel().rows.length;

  return (
    <div
      className="flex items-center justify-between px-2 py-4"
     
    >
      {/* 左側：頁面大小選擇器 */}
      <div className="flex items-center space-x-2">
        <p className="text-sm font-medium">
          每頁顯示
        </p>
        <Select
          value={`${pageSize}`}
          onValueChange={(value) => {
            table.setPageSize(Number(value));
          }}
         
        >
          <SelectTrigger className="h-8 w-[70px]">
            <SelectValue placeholder={pageSize} />
          </SelectTrigger>
          <SelectContent side="top">
            {pageSizeOptions.map((pageSize) => (
              <SelectItem
                key={pageSize}
                value={`${pageSize}`}
               
              >
                {pageSize}
              </SelectItem>
            ))}
          </SelectContent>
        </Select>
        <p className="text-sm font-medium">
          筆
        </p>
      </div>

      {/* 中間：數據範圍顯示 */}
      <div
        className="flex items-center justify-center text-sm font-medium"
       
      >
        顯示第 {startRow} - {endRow} 筆，共 {totalRows} 筆
      </div>

      {/* 右側：分頁導航 */}
      <div className="flex items-center space-x-2">
        {/* 頁面信息 */}
        <div className="flex items-center space-x-2">
          <p className="text-sm font-medium">
            第 {currentPage} 頁，共 {totalPages} 頁
          </p>
        </div>

        {/* 分頁按鈕 */}
        <div className="flex items-center space-x-1">
          {/* 第一頁 */}
          <Button
            variant="outline"
            className="hidden h-8 w-8 p-0 lg:flex"
            onClick={() => table.setPageIndex(0)}
            disabled={!table.getCanPreviousPage()}
            aria-label="跳到第一頁"
           
          >
            <ChevronsLeftIcon className="h-4 w-4" />
          </Button>

          {/* 上一頁 */}
          <Button
            variant="outline"
            className="h-8 w-8 p-0"
            onClick={() => table.previousPage()}
            disabled={!table.getCanPreviousPage()}
            aria-label="上一頁"
           
          >
            <ChevronLeftIcon className="h-4 w-4" />
          </Button>

          {/* 下一頁 */}
          <Button
            variant="outline"
            className="h-8 w-8 p-0"
            onClick={() => table.nextPage()}
            disabled={!table.getCanNextPage()}
            aria-label="下一頁"
           
          >
            <ChevronRightIcon className="h-4 w-4" />
          </Button>

          {/* 最後一頁 */}
          <Button
            variant="outline"
            className="hidden h-8 w-8 p-0 lg:flex"
            onClick={() => table.setPageIndex(table.getPageCount() - 1)}
            disabled={!table.getCanNextPage()}
            aria-label="跳到最後一頁"
           
          >
            <ChevronsRightIcon className="h-4 w-4" />
          </Button>
        </div>
      </div>
    </div>
  );
}
