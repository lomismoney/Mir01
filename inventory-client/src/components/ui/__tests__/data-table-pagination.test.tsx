import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DataTablePagination } from "../data-table-pagination";

/**
 * DataTablePagination 組件測試
 * 
 * 測試範圍：
 * - 組件正確渲染
 * - 分頁資訊顯示
 * - 分頁按鈕功能
 * - 頁面大小選擇
 * - 按鈕狀態管理
 * - 邊界情況處理
 */

// 創建模擬的 table 實例
const createMockTable = (pageIndex = 0, pageSize = 10, pageCount = 5, canPreviousPage = true, canNextPage = true) => {
  return {
    getState: () => ({
      pagination: {
        pageIndex,
        pageSize,
      },
    }),
    getPageCount: () => pageCount,
    getFilteredRowModel: () => ({
      rows: Array.from({ length: pageSize }, (_, i) => ({ id: i })),
    }),
    getCanPreviousPage: () => canPreviousPage,
    getCanNextPage: () => canNextPage,
    setPageSize: jest.fn(),
    setPageIndex: jest.fn(),
    previousPage: jest.fn(),
    nextPage: jest.fn(),
  } as any;
};

describe("DataTablePagination", () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    user = userEvent.setup();
  });

  /**
   * 測試基本渲染功能
   */
  it("應該正確渲染分頁組件", () => {
    const mockTable = createMockTable();
    render(<DataTablePagination table={mockTable} />);
    
    // 檢查分頁文字存在
    expect(screen.getByText("每頁顯示")).toBeInTheDocument();
    expect(screen.getByText("筆")).toBeInTheDocument();
    expect(screen.getByText(/顯示第/)).toBeInTheDocument();
    expect(screen.getByText(/第 1 頁，共 5 頁/)).toBeInTheDocument();
  });

  /**
   * 測試數據範圍顯示
   */
  it("應該正確顯示數據範圍", () => {
    const mockTable = createMockTable(0, 10, 5);
    render(<DataTablePagination table={mockTable} totalCount={50} />);
    
    // 檢查數據範圍顯示
    expect(screen.getByText("顯示第 1 - 10 筆，共 50 筆")).toBeInTheDocument();
  });

  /**
   * 測試第二頁的數據範圍顯示
   */
  it("應該正確顯示第二頁的數據範圍", () => {
    const mockTable = createMockTable(1, 10, 5);
    render(<DataTablePagination table={mockTable} totalCount={50} />);
    
    // 檢查第二頁的數據範圍
    expect(screen.getByText("顯示第 11 - 20 筆，共 50 筆")).toBeInTheDocument();
    expect(screen.getByText("第 2 頁，共 5 頁")).toBeInTheDocument();
  });

  /**
   * 測試分頁按鈕存在性
   */
  it("應該顯示所有分頁按鈕", () => {
    const mockTable = createMockTable(1, 10, 5);
    render(<DataTablePagination table={mockTable} />);
    
    // 檢查所有分頁按鈕
    expect(screen.getByLabelText("跳到第一頁")).toBeInTheDocument();
    expect(screen.getByLabelText("上一頁")).toBeInTheDocument();
    expect(screen.getByLabelText("下一頁")).toBeInTheDocument();
    expect(screen.getByLabelText("跳到最後一頁")).toBeInTheDocument();
  });

  /**
   * 測試第一頁時按鈕禁用狀態
   */
  it("在第一頁時應該禁用前一頁和第一頁按鈕", () => {
    const mockTable = createMockTable(0, 10, 5, false, true);
    render(<DataTablePagination table={mockTable} />);
    
    // 檢查按鈕禁用狀態
    expect(screen.getByLabelText("跳到第一頁")).toBeDisabled();
    expect(screen.getByLabelText("上一頁")).toBeDisabled();
    expect(screen.getByLabelText("下一頁")).not.toBeDisabled();
    expect(screen.getByLabelText("跳到最後一頁")).not.toBeDisabled();
  });

  /**
   * 測試最後一頁時按鈕禁用狀態
   */
  it("在最後一頁時應該禁用下一頁和最後一頁按鈕", () => {
    const mockTable = createMockTable(4, 10, 5, true, false);
    render(<DataTablePagination table={mockTable} />);
    
    // 檢查按鈕禁用狀態
    expect(screen.getByLabelText("跳到第一頁")).not.toBeDisabled();
    expect(screen.getByLabelText("上一頁")).not.toBeDisabled();
    expect(screen.getByLabelText("下一頁")).toBeDisabled();
    expect(screen.getByLabelText("跳到最後一頁")).toBeDisabled();
  });

  /**
   * 測試上一頁按鈕點擊
   */
  it("點擊上一頁按鈕應該調用 previousPage", async () => {
    const mockTable = createMockTable(1, 10, 5);
    render(<DataTablePagination table={mockTable} />);
    
    const prevButton = screen.getByLabelText("上一頁");
    await user.click(prevButton);
    
    expect(mockTable.previousPage).toHaveBeenCalled();
  });

  /**
   * 測試下一頁按鈕點擊
   */
  it("點擊下一頁按鈕應該調用 nextPage", async () => {
    const mockTable = createMockTable(1, 10, 5);
    render(<DataTablePagination table={mockTable} />);
    
    const nextButton = screen.getByLabelText("下一頁");
    await user.click(nextButton);
    
    expect(mockTable.nextPage).toHaveBeenCalled();
  });

  /**
   * 測試跳到第一頁按鈕點擊
   */
  it("點擊跳到第一頁按鈕應該調用 setPageIndex(0)", async () => {
    const mockTable = createMockTable(2, 10, 5);
    render(<DataTablePagination table={mockTable} />);
    
    const firstButton = screen.getByLabelText("跳到第一頁");
    await user.click(firstButton);
    
    expect(mockTable.setPageIndex).toHaveBeenCalledWith(0);
  });

  /**
   * 測試跳到最後一頁按鈕點擊
   */
  it("點擊跳到最後一頁按鈕應該調用 setPageIndex(pageCount - 1)", async () => {
    const mockTable = createMockTable(1, 10, 5);
    render(<DataTablePagination table={mockTable} />);
    
    const lastButton = screen.getByLabelText("跳到最後一頁");
    await user.click(lastButton);
    
    expect(mockTable.setPageIndex).toHaveBeenCalledWith(4);
  });

  /**
   * 測試頁面大小選擇器
   */
  it("應該正確顯示頁面大小選擇器", () => {
    const mockTable = createMockTable();
    render(<DataTablePagination table={mockTable} />);
    
    // 檢查當前頁面大小顯示（在 select 觸發器中）
    expect(screen.getByText("10")).toBeInTheDocument();
    expect(screen.getByRole("combobox")).toBeInTheDocument();
  });

  /**
   * 測試自定義頁面大小選項
   */
  it("應該支援自定義頁面大小選項", async () => {
    const mockTable = createMockTable();
    const customOptions = [5, 10, 25, 50];
    
    render(
      <DataTablePagination 
        table={mockTable} 
        pageSizeOptions={customOptions} 
      />
    );
    
    // 點擊選擇器打開下拉選單
    const trigger = screen.getByRole("combobox");
    await user.click(trigger);
    
    // 檢查自定義選項是否存在
    expect(screen.getByText("5")).toBeInTheDocument();
    expect(screen.getByText("25")).toBeInTheDocument();
    expect(screen.getByText("50")).toBeInTheDocument();
  });

  /**
   * 測試頁面大小變更
   */
  it("變更頁面大小應該調用 setPageSize", async () => {
    const mockTable = createMockTable();
    render(<DataTablePagination table={mockTable} />);
    
    // 點擊選擇器
    const trigger = screen.getByRole("combobox");
    await user.click(trigger);
    
    // 選擇新的頁面大小
    const option20 = screen.getByText("20");
    await user.click(option20);
    
    expect(mockTable.setPageSize).toHaveBeenCalledWith(20);
  });

  /**
   * 測試無 totalCount 時的行為
   */
  it("沒有 totalCount 時應該使用 filtered rows 長度", () => {
    const mockTable = createMockTable();
    render(<DataTablePagination table={mockTable} />);
    
    // 應該顯示 filtered rows 的數量
    expect(screen.getByText("顯示第 1 - 10 筆，共 10 筆")).toBeInTheDocument();
  });

  /**
   * 測試組件快照
   */
  it("應該匹配快照", () => {
    const mockTable = createMockTable();
    const { container } = render(<DataTablePagination table={mockTable} />);
    expect(container.firstChild).toMatchSnapshot();
  });
}); 