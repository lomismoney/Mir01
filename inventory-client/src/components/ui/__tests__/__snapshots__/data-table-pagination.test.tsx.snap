// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTablePagination 應該匹配快照 1`] = `
<div
  class="flex items-center justify-between px-2 py-4"
>
  <div
    class="flex items-center space-x-2"
  >
    <p
      class="text-sm font-medium"
    >
      每頁顯示
    </p>
    <button
      aria-autocomplete="none"
      aria-controls="radix-«r3g»"
      aria-expanded="false"
      class="border-input data-[placeholder]:text-muted-foreground [&_svg:not([class*='text-'])]:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 dark:hover:bg-input/50 flex items-center justify-between gap-2 rounded-md border bg-transparent px-3 py-2 text-sm whitespace-nowrap shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 data-[size=default]:h-9 data-[size=sm]:h-8 *:data-[slot=select-value]:line-clamp-1 *:data-[slot=select-value]:flex *:data-[slot=select-value]:items-center *:data-[slot=select-value]:gap-2 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 h-8 w-[70px]"
      data-size="default"
      data-slot="select-trigger"
      data-state="closed"
      dir="ltr"
      role="combobox"
      type="button"
    >
      <span
        data-slot="select-value"
        style="pointer-events: none;"
      >
        10
      </span>
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down size-4 opacity-50"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </button>
    <p
      class="text-sm font-medium"
    >
      筆
    </p>
  </div>
  <div
    class="flex items-center justify-center text-sm font-medium"
  >
    顯示第 
    1
     - 
    10
     筆，共 
    10
     筆
  </div>
  <div
    class="flex items-center space-x-2"
  >
    <div
      class="flex items-center space-x-2"
    >
      <p
        class="text-sm font-medium"
      >
        第 
        1
         頁，共 
        5
         頁
      </p>
    </div>
    <div
      class="flex items-center space-x-1"
    >
      <button
        aria-label="跳到第一頁"
        class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 has-[>svg]:px-3 hidden h-8 w-8 p-0 lg:flex"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevrons-left h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m11 17-5-5 5-5"
          />
          <path
            d="m18 17-5-5 5-5"
          />
        </svg>
      </button>
      <button
        aria-label="上一頁"
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 has-[>svg]:px-3 h-8 w-8 p-0"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-left h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m15 18-6-6 6-6"
          />
        </svg>
      </button>
      <button
        aria-label="下一頁"
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 has-[>svg]:px-3 h-8 w-8 p-0"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </button>
      <button
        aria-label="跳到最後一頁"
        class="items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 has-[>svg]:px-3 hidden h-8 w-8 p-0 lg:flex"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevrons-right h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 17 5-5-5-5"
          />
          <path
            d="m13 17 5-5-5-5"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
